concordium-consensus/src/Concordium/GlobalState/Persistent/Cache.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Concordium.GlobalState.Persistent.Cache where
+
+import Control.Concurrent.MVar
+import Control.Monad.Except (ExceptT)
+import Control.Monad.IO.Class
+import Control.Monad.Reader
+import Control.Monad.State.Strict
+import Control.Monad.Writer.Strict (WriterT)
+import qualified Data.IntMap.Strict as IntMap
+import Data.Proxy
+import qualified Data.Vector.Primitive.Mutable as Vec
+
+import Concordium.Utils.Serialization.Put (PutT)
+
+import Concordium.GlobalState.Persistent.BlobStore (BlobRef (..), BlobStoreT, HasNull (..))
+
+-- |The 'HasCache' class supports projecting a cache of a particular type.
+-- This is used to access a cache given a context.
+class HasCache cache r where
+    -- |Project the cache.
+    projectCache :: r -> cache
+
+-- |The 'MonadCache' combines 'MonadIO' with an operation for getting the cache.
+-- This is used as a constraint in the 'Cache' class.
+class MonadIO m => MonadCache c m where
+    -- |Get the cache.
+    getCache :: m c
+    default getCache :: (MonadReader r m, HasCache c r) => m c
+    getCache = asks projectCache
+    {-# INLINE getCache #-}
+
+instance MonadCache c m => MonadCache c (StateT s m) where
+    getCache = lift getCache
+    {-# INLINE getCache #-}
+
+instance (Monoid w, MonadCache c m) => MonadCache c (WriterT w m) where
+    getCache = lift getCache
+    {-# INLINE getCache #-}
+
+instance (MonadCache c m) => MonadCache c (ExceptT e m) where
+    getCache = lift getCache
+    {-# INLINE getCache #-}
```

Comment:
I am not sure what other behaviour would be considered natural. I assume you mean some kind of rollback of the cache in the error case, but I do not think this would be reasonable or natural.